### ISSUE TITLE
chrome apps and database csp

### ISSUE THREAD

**Issue by agosto-chrisdavis** [firebase/firebase-js-sdk#82]
* Platform: Chrome App 
  * Operating System: Ubuntu and Chrome OS
  * Firebase SDK version: 4.1.3
  * Firebase Product: database

When I run the sdk in a chrome app, I get csp errors when using the database. I see you have an example for chrome extensions (https://github.com/firebase/quickstart-js/tree/master/auth/chromextension), but since you can't set the content_security_policy in an chrome app's manifest.json, that example won't work for apps.

Here's an example of the error on stackoverflow: https://stackoverflow.com/questions/29399594/error-using-firebase-from-chrome-app

```
Refused to load the script 'https://{my firebase id}.firebaseio.com/.lp?start=t&ser=81980096&cb=15&v=5' because it violates the following Content Security Policy directive: "default-src 'self' chrome-extension-resource:". Note that 'script-src' was not explicitly set, so 'default-src' is used as a fallback.
```

I want to solve the above issue but I also want to know why it's only using the https long_polling method for the database and not the websockets wss, which I normally see in my traditional web app. I don't see any reason you can't use wss in a chrome app and it might solve the above issue. It appears the sdk is not even trying to use websockets and just defaults to long polling.

It also seem you're using window.open somewhere which is also causing issues since it's not available in chrome apps.

In short, how can I use the sdk in a chrome app?

```javascript
//  using custom token
firebase.initializeApp(config);
firebase.auth().signInWithCustomToken(token);

// later on
firebase.database()
        .ref(databaseRefPath())
        .set(getRepresentation())
        .catch(e=>this.log.w(`persist error: ${e}`));
```

**Comment by mikelehen** [comment]
Hi!  Sorry you're running into issues with chrome apps. :-(

If you're interested, the code for choosing what transport(s) to use lives here: https://github.com/firebase/firebase-js-sdk/blob/master/src/database/js-client/realtime/TransportManager.js#L49

The basic approach is: 

1. Try WebSockets first.
2. If WebSockets fail, fall back to long-polling and remember that they failed so we don't waste time trying WebSockets every time you load the page.
3. If long-polling succeeds, attempt to "upgrade" to WebSockets in the background.

We do this because WebSockets is more efficient but (historically at least), long-polling is more reliable in more browser environments.  So we end up preferring Long-Polling and only once that succeeds, try WebSockets again.  Chrome apps are apparently an exception where long-polling should be outright disabled.

The proper fix for this would probably be to detect chrome apps and disable long-polling support here: https://github.com/firebase/firebase-js-sdk/blob/master/src/database/js-client/realtime/BrowserPollConnection.js#L202

But as a temporary solution, you could probably move forward by changing your databaseURL in your firebase.initalializeApp() call to be "wss://..." instead of "https://..."   This is a trick to tell the client to only use WebSockets.

Hope this helps.  If you're interested in providing a PR to fix the core issue and want any additional pointers, let us know!

**Comment by agosto-chrisdavis** [comment]
Using `wss:// ` worked for getting the database working.  Thanks for the tip.  

Is that documented somewhere?  I was looking for a way to force websockets but failed. 

I'll see about providing a PR to fix this one issue.  

I get one other error in the logs when running in a chrome app, but doesn't see to be preventing me from using the sdk so far.

`window.localStorage is not available in packaged apps. Use chrome.storage.local instead.`

**Comment by mikelehen** [comment]
wss:// isn't really documented since it's not supposed to be necessary and generally shouldn't be used.  I've written a self-answered Stack Overflow question though so future people may find it if necessary: https://stackoverflow.com/questions/44937825/how-can-i-force-firebase-realtime-database-to-use-websockets/44937826#44937826

I think the SDK optimistically tries to use localStorage for a few things but gracefully handles it if it's not available, so that error should be benign.  You could check the source code on github to find the specific places it's used.

**Comment by PierBover** [comment]
I also had to resort to using `wss://` to avoid de CSP error in Chrome OS.

> I think the SDK optimistically tries to use localStorage for a few things but gracefully handles it if it's not available, so that error should be benign. You could check the source code on github to find the specific places it's used.

The error doesn't affect the performance of the app, but when closing and reopening the app the session is not maintained. So the SDK is not really handling it gracefully. I would expect the JWT to be stored somewhere else.

I'm thinking I could store the token myself using `chrome.storage.local` or something along those lines.

How did you solve that @agosto-chrisdavis ?

Another solution would be to switch back to the REST API and just use Axios... but quite frankly that would be pretty ugly. Even more considering Firebase is a Google product and so is Chrome OS.

**Comment by bojeil-google** [comment]
Hey @PierBover, I only tested with Chrome extensions and the Auth SDK should work there. Now that you mention, localStorage is not supported in Chrome apps. We are probably falling back to in memory state storage. We have plans to switch to `indexedDB`. I prefer that over `chrome.storage.local` as it is available by default and works in majority of supported browsers including extensions, service/web worker environments, Ionic/Cordova, etc and doesn't require permission. However, that will take time to implement.

**Comment by bojeil-google** [comment]
One hack you can do in the meantime, is create your own polyfill of localStorage which you persist underneath with `chrome.storage.local`. On page reload, you copy that content back to your polyfill before initializing the auth instance.

**Comment by PierBover** [comment]
Here is a small polyfill I wrote for `localStorage` using `chrome.storage.local`.

https://github.com/PierBover/chrome-os-localstorage-polyfill

Hopefully someone will find it useful.
